File: medium/2069_walking_robot_simulation_ii/solution.cpp
#include <string>
#include <vector>

class Robot {
private:
  int width, height;
  int perimeter;

  int x = 0;
  int y = 0;

  int direction = 0;
  const std::string directions[4] = {"East", "North", "West", "South"};

public:
  Robot(int width, int height) : width(width), height(height) {
    this->perimeter = 2 * (width + height) - 4;
  }

  void step(int num) {
    if (this->perimeter == 0)
      return;

    num %= this->perimeter;
    if (num == 0)
      num = this->perimeter;

    int distance;

    while (num > 0) {
      if (this->direction == 0) {
        distance = (this->width - 1) - this->x;
        if (num <= distance) {
          this->x += num;
          return;
        }
        this->x = this->width - 1;
      } else if (this->direction == 1) {
        distance = (this->height - 1) - this->y;
        if (num <= distance) {
          this->y += num;
          return;
        }
        this->y = this->height - 1;
      } else if (this->direction == 2) {
        distance = this->x;
        if (num <= distance) {
          this->x -= num;
          return;
        }
        this->x = 0;
      } else {
        distance = this->y;
        if (num <= distance) {
          this->y -= num;
          return;
        }
        this->y = 0;
      }

      num -= distance;
      this->direction = (this->direction + 1) % 4;
    }
  }

  std::vector<int> getPos() { return std::vector<int>{this->x, this->y}; }

  std::string getDir() { return this->directions[this->direction]; }
};

File: medium/2069_walking_robot_simulation_ii/solution.py
from typing import List


class Robot:
    directions = {0: "North", 90: "East", 180: "South", 270: "West"}

    def __init__(self, width: int, height: int):
        self.width = width
        self.height = height

        self.x = 0
        self.y = 0
        self.direction = 90

    def step(self, num: int) -> None:
        if self.direction == 0:
            self.y += num
        elif self.direction == 90:
            self.x += num
        elif self.direction == 180:
            self.y -= num
        elif self.direction == 270:
            self.x -= num

        if self.x < 0:
            num = abs(self.x)
            self.x = 0
        elif self.x >= self.width:
            num = self.x - (self.width - 1)
            self.x -= num
        elif self.y < 0:
            num = abs(self.y)
            self.y = 0
        elif self.y >= self.height:
            num = self.y - (self.height - 1)
            self.y -= num
        else:
            num = 0

        if num:
            self.direction = (self.direction - 90) % 360
            self.step(num)

    def getPos(self) -> List[int]:
        return [self.x, self.y]

    def getDir(self) -> str:
        return self.directions[self.direction]
